cfe2cos, the mega-merge

Others opened the pull request:
### Summary of this Pull Request (PR)
The cFE is a complex piece of software, consequently our support implementation is quite complex. I will attempt to summarize what is going on with each change.

#### cFE_booter Support Code 
- Summary: To get from `cos_init` to the actual cFE is a bit of code, most of which is inspired by the posix implementation. We need to set up all of our data structures, start our scheduler, and make sure the filesystem is in good order. It's spread across of few files -- but again, not too complicated. We also boost the stack size so that the cFE doesn't blow the small stack. You can trace configuration logic from `cFE_entrypoint.c` The only other interesting stuff is in `cFE_util.c`.
- Component files: `cFE_booter/cFE_entrypoint.c` `cFE_booter/cFE_util.c` `cFE_booter/cFE_util.h` parts of `cFE_booter/osapi.c` and `cFE_booter/ostask.c` `include/shared/consts.h`
- Known Issues: The first code we wrote. It's a bit overly complex and stupid. It's written in a very formal style, which we quickly diverged from.

#### General OSAL Method Implementations
- Summary: The most straightforward of the straightforward. Some of these are a bit hacky, but the code quality is fine. Many of these methods are irrelevant and unimplemented.
- Component files: `cFE_booter/osapi.c` `cFE_booter/psp.c` `cFE_booter/ostimer.c` `cFE_booter/osnetwork.c`
- Known Issues:

#### Copied Files
- Summary: Shared OSAL code we just copied out of the cFE. Unmodified, doesn't need review.
- Component files: `cFE_booter/cfe_psp_*`
- Known Issues:

#### App Loading
- Summary: Code to work as glue between how the cFE thinks of apps, and how composite thinks of components. Method glue code found elsewhere. 
- Component files: `cFE_booter/osloader.c`
- Known Issues: Creating a new app is a bit of a janky process.  

#### cFE Tasks + Scheduling
- Summary: My baby. Glue for making the cFE task API compatible with composite threads. Relies on a lot of sl code under the hood. Probably needs checking for races. 
- Component files: `cFE_booter/ostask.c` `cFE_booter/ostask.h` `cFE_booter/sl_mod_fprr.c` `cFE_booter/sl_mod_policy.h` `cFE_booter/sl_thd_static_backend.c`
- Known Issues: The semaphore implementation is pure garbage. (Luckily they're not really used.) We should talk about zero allocation semaphores.

#### OSAL Filesystem APIS
- Summary: This is all Joe's doing, he'll leave a comment on it.
- Component files:  `cFE_booter/osfiles.c` `cFE_booter/osfilesys.c` `cFE_booter/osfilesys.h` `cFE_booter/tar.c` `cFE_booter/tar.h`
- Known Issues:

#### Queue Implementation
- Summary: Zach's doing. Not much interesting here, and I haven't looked at it in a long time. Appears to work fine.
- Component files: `cFE_booter/osqueue.c`
- Known Issues:

#### Major Build System Changes
- Summary: Integrating the cFE under composite took a bit of `make` wrangling. We kept most of our original build logic--in python--but it now runs totally under composite's `make`. I think most of the custom stuff is well commented.
- Component files: `.gitignore` `.gitmodules` `src/Makefile` `extern/make.py`
- Known Issues: Half of it is in python. My `make`foo is weak, so some of this is a bit of a cludge. I just don't know how to do it better.

#### cFE App Component Wrappers
- Summary: We wrap every cFE app in a component, and each of these components needs a little bit of helper code. Lots of mostly unavoidable code duplication.
- Component files: `no_interface/sch_lab/*` `no_interface/sample_lib/*` `no_interface/sample_app/*`
- Known Issues: 

#### cFE emulation for apps
- Summary: We need to provide a way for apps to call into the cFE, but intercepting the calls and writing parts into shared memory.
- Component files: `src/components/Makefile.comp` `cFE_booter/cFE_emu_support.c` `lib/cFE_emu.c` `lib/Makefile` `include/cFE_emu.h`
- Known Issues: I've just realized that this should've been done through c stubs. I'll fix that, so don't bother commenting about it.

#### cFE_booter Build System
- Summary: The cFE booter needs some some special build logic. It isn't anything too complicated, just integrating a few external object files.
- Component files: `cFE_booter/.gitignore` `cFE_booter/Makefile` 
- Known Issues:

#### Running the cFE
- Summary: These are our runscripts and Vagrantfile, how the cFE actually gets going. Not that interesting or important.
- Component files: `tools/Vagrantfile` `runscripts/cFE_booter.sh` `runscripts/llboot_cFE.sh`
- Known Issues:


#### Other Outstanding Issues  
- The cFE unit tests are currently broken. This probably should be fixed, but I'm unsure how much work it is. Likely can be done after the deadline.
- `llbooter/boot_deps.h` and `llbooter/llbooter.c` are modified to map an extra page per component. This sucks, but without it the cFE breaks.
- A bunch of stub files are changed. (As well as `include/cos_asm_simple_stacks.h`) This is because their composite versions are utterly broken and don't respect `consts.h`. I don't implement the most elegant solution, but w/e.
 
### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [x] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
(Specify @<github.com username(s)> of the reviewers. Ex: @user1, @user2)
@gparmer @base0x10 @phanikishoreg 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [ ] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [x] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing
I've tested the code using the following test programs (provide list here):

- the cFE

Others commented:
@gparmer Tried my best to summarize everything. Let me know if I missed anything important out.

gparmer commented:
@RobertGiff You should likely have a look at the build system stuff to see if it is compatible with the RK stuff.

gparmer commented on src/extern/make.py:
```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python2.7
```
yeah, this file is pretty....unique.  I'm fine with it for now, but it should not be in the `extern/` dir.  See below.

gparmer commented on src/Makefile:
```diff
@@ -1,9 +1,27 @@
+	$(info ********** Setting up External Libs ***********)
+	$(info ***********************************************)
+	$(info )
+	bash -c 'cd extern/cFE/; source setvars.sh; cd ../..; make -C extern/cFE/build/cpu1 config'
```
We can hide all the python specific stuff inside a `cFS` subdir.  See comments later.

phanikishoreg commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 	url = https://github.com/gwsystems/ps.git
+[submodule "src/extern/cFE"]
+	path = src/extern/cFE
+	url = https://github.com/Others/cFE.git
```
Should this be sourced from gwsystems instead?

Others replied:
No. This is not the cFE repository on gwsystems. (Which really should be changed back to `cFE2cos`...) This cFE repo just has cFE code in it.

phanikishoreg commented on src/Makefile:
```diff
@@ -36,8 +54,12 @@ distclean: clean
 	@rm -f `pwd`/../transfer/*
 
-init:
+init: | ensure_config init_extern init_composite
```
Never used `|` in such a rule before. I'll need to look up what that does.

Others replied:
It sequences the dependencies

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/.gitignore:
```diff
@@ -0,0 +1,3 @@
+gen
```
why another .gitignore in the same git repo?

Others replied:
It's more flexible. We don't need super long paths in the top level `.gitignore`

phanikishoreg replied:
I see. I'm not sure how this works, does git ignore files from both local .gitignore and the global repository .gitignore? 
Also, this is probably the longest path we can have in our build system (in terms of levels, not names), so it's not super-long if you think of it.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/Makefile:
```diff
@@ -0,0 +1,12 @@
+INTERFACES=cFE
+DEPENDENCIES=capmgr
+IF_LIB:=./composite_cFE.o ./cFE_fs.o $(wildcard test/*.o)
+ADDITIONAL_LIBS=-lcobj_format -lcos_kernel_api -lcos_defkernel_api -lsl_capmgr -lsl_sched -lheap -lsl_lock
```
I've added a macro for `-lcos_kernel_api -lcos_defkernel_api -lsl_capmgr -lsl_sched` = LIBSLCAPMGR, you can probably use that.

Others replied:
Awesome! That's very convenient

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+#include <cfe_error.h>
+
+#include <cFE_emu.h>
+
```
Too many blank lines in header inclusion.

Others replied:
We follow the std => cos => interface => cFE => internal header convention

phanikishoreg replied:
It would be better to have that as comments, so it's followed by others who change these files. 
Like,
```
/* standard includes */
xxx
...
/* composite includes */
...
/* interface includes */
...
/* cFE includes */
...
```
These undocumented groupings may not be followed otherwise.

Others replied:
It's just a convention we follow. Plus, that's like 4 comments on headers per file. That seems pretty extra, since most of composite groups headers similarly without comments.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+{
+	vaddr_t our_addr = 0;
+	int     id       = memmgr_shared_page_alloc(&our_addr);
+	assert(our_addr);
```
style, separation of declarations | body | return.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+union shared_region *shared_regions[16];
+
+int
+emu_backend_request_memory(spdid_t client)
```
should it be `cFE_emu`?

Others replied:
We're already in the `cFE` component. No need to namespace unnecessarily IMHO. I'm open to changing this tho.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/cfe_psp_eeprom.c:
```diff
@@ -0,0 +1,352 @@
+/*
```
Did you have to make changes in this file? If not, perhaps just a soft-link from the extern directory would be better.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/cfe_psp_memrange.c:
```diff
@@ -0,0 +1,311 @@
+/*
```
ditto

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,794 @@
+void
+OS_IdleLoop(void)
+{
+	while (1) sl_thd_block(0);
```
should be `while (1) sl_thd_yield();` instead so if there is nothing else to run, this runs instead of blocking.
Anyone wakes this up?

Others replied:
This is called from the root thread, which has a super high priority. Your suggestion would imply we'd need to go through a yield every time we schedule a thread. (Even though that thread would never do anything but yield!)

phanikishoreg replied:
My bad, I missed the comment before that function. Yeah, yield is not the right option here.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/sl_mod_fprr.c:
```diff
@@ -0,0 +1,120 @@
+#include <llprint.h>
```
I forget why this is here and what the difference is between this and `components/lib/sl/sl_mod_fprr.c`.

phanikishoreg replied:
If it's identical, perhaps soft-link.

Others replied:
It's not identical, we need more priorities.

phanikishoreg commented on src/components/implementation/no_interface/cFE_booter/sl_thd_static_backend.c:
```diff
@@ -0,0 +1,69 @@
+/**
```
yeah, code duplication. use soft-links instead if this is necessary.

phanikishoreg commented on tools/Vagrantfile:
```diff
@@ -1,19 +1,30 @@
 # -*- mode: ruby -*-
```
`tools` in composite tree? or should it be part of `cFE` tree in extern?

Others replied:
It works for non cFE composite development. Open to changing this though.

Others commented:
I've addressed most of the above comments, will investigate the others. I also got rid of the cFE_emu library, and used the existing c_stubs mechanism.

@gparmer In terms of review ordering, this order would be great:

- cFE_booter Support Code
- cFE Tasks + Scheduling
- General OSAL Method Implementations
- Queue Implementation

Otherwise the original order of the summaries is a proxy for my desire to have that part reviewed. (Although, I think the filesystem should go last. It's already been reviewed after all.)

Thanks!

phanikishoreg commented on src/components/interface/cFE/cFE_emu.h:
```diff
@@ -0,0 +1,106 @@
+
+int emu_backend_request_memory(spdid_t client);
+
+int32 emu_CFE_ES_GetGenCount(spdid_t client);
```
You should use the `cos_inv_token()` in the server component for the client's spdid instead.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,793 @@
+	struct sl_thd *         main_delegate_thread = sl_thd_alloc(main_delegate, NULL);
+	union sched_param_union sp = {.c = {.type = SCHEDP_PRIO, .value = MAIN_DELEGATE_THREAD_PRIORITY}};
+	sl_thd_param_set(main_delegate_thread, sp.v);
+	main_delegate_thread_id = sl_thd_thdid(main_delegate_thread);
```
"delegate thread" is a little bit of a strange name.  Not really sure what "delegate" has to do with all of this.

Others replied:
It's called `main_delegate_thread`. It's the thread we delegate main execution to. Do you have a better name in mind?

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,793 @@
+			                           it. */
+		}
+
+		sl_thd_block_periodic(0);
```
just to make sure we aren't spinning in this loop for some strange interaction in the system, it might be smart to count how many times we've iterated without triggering the `now > start` condition, and assert if it goes above some level (say 64).

Others commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+#include <cos_component.h>
+#include <cos_debug.h>
+#include <cos_types.h>
+
```
We follow the std => cos => interface => cFE => internal header convention. Otherwise it becomes very confusing where each header is coming from. Especially since we have 5 header sources. Is this something the cos style guide takes a strong stance on?

Others commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+emu_CFE_SB_InitMsg(spdid_t client)
+{
+	union shared_region *s = shared_regions[client];
+	CFE_SB_InitMsg(s->cfe_sb_initMsg.MsgBuffer, s->cfe_sb_initMsg.MsgId, s->cfe_sb_initMsg.Length,
```
Can you explain what you mean by that? Isn't `interface/cfe/` for the marshaling functions on the client side?

gparmer replied:
The functions on the client and server side of an invocation through an interface function should be in sstubs.c and cstubs.c in the interface.  I'm not sure if that is what these are doing.

Others commented on src/components/implementation/no_interface/cFE_booter/osapi.c:
```diff
@@ -0,0 +1,457 @@
+int have_initialized = 0;
+
+int32
+OS_API_Init(void)
```
It's that way because the OSAL unit tests expect this to be idempotent. We can't duplicate initialization, so I'm not sure what a better solution would look like.

Others commented on src/components/implementation/no_interface/cFE_booter/osapi.c:
```diff
@@ -0,0 +1,457 @@
+	microsec_t old_seconds         = (microsec_t)initial_time.seconds;
+	microsec_t old_additional_usec = (microsec_t)initial_time.microsecs;
+
+	microsec_t old_usec = old_seconds * (1000 * 1000) + old_additional_usec;
```
It's 64 bit.

Others commented on src/components/implementation/no_interface/cFE_booter/osapi.c:
```diff
@@ -0,0 +1,457 @@
+
+	switch (error_num) {
+	case OS_SUCCESS:
+		strcpy(local_name, "OS_SUCCESS");
```
We need to return into a `os_err_name_t`. This is literally copied from the posix implementation, and I don't see a good reason to change it.

gparmer replied:
I don't understand either of these justifications.  `*err_name` is a `os_err_name_t`.  If posix did something dumb I don't see a reason to copy it.  There might be other arguments; these seems lazy.  I don't have time to go through all comments to see if they are making good justifications or not, so this isn't heartening.

Others commented on src/components/implementation/no_interface/cFE_booter/cFE_util.c:
```diff
@@ -0,0 +1,32 @@
+void
+print_with_error_name(char *message, int32 error)
+{
+	os_err_name_t local_name;
```
This is `cFE` core stuff. There is no reasonable way for me to fix this without adding a dependency on a cFE implementation detail.

Others commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,793 @@
+	cycles_t now, start;
+
+	rdtscll(start);
+	start += sl_usec2cyc(HZ_PAUSE);
```
Uniformly 64 bit.

Others commented on src/components/implementation/no_interface/cFE_booter/cFE_emu_support.c:
```diff
@@ -0,0 +1,117 @@
+}
+
+void
+emu_CFE_SB_InitMsg(spdid_t client)
```
No. I didn't think they should be?

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+	union sched_param_union sp = {.c = {.type = SCHEDP_PRIO, .value = priority}};
+	sl_thd_param_set(thd, sp.v);
+
+	struct cfe_task_info *task_info = &cfe_tasks[sl_thd_thdid(thd)];
```
ha, our thdid function has a pretty redundant name.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+	struct sl_thd *thd = sl_thd_lkup(task_id);
+	if (!thd) { return OS_ERR_INVALID_ID; }
+
+	struct cfe_task_info *task_info = &cfe_tasks[task_id];
```
def/blocketc

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+OS_TaskGetId(void)
+{
+	thdid_t real_id           = sl_thdid();
+	thdid_t possible_override = id_overrides[real_id];
```
interesting.  not sure where this is set (will grep through), but I'm hoping there is a comment.

gparmer replied:
OK, I understand what it is doing, but it needs a comment when it is set in `launch_other_component`.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+{
+	if (OS_TaskGetId() == main_delegate_thread_id) { return OS_ERR_INVALID_ID; }
+
+	struct cfe_task_info *task_info = &cfe_tasks[sl_thd_thdid(sl_thd_curr())];
```
The whole `cfe_tasks` could/should be replaced by our own thread policy structure for cfe, right?  Are there TODOs for this, and will we add issues after merge?  I know that you want to wait to solve this properlly, if I rememer correctly.

Others replied:
I really didn't want to spend time fixing sl, when we're not even sure what the design should look like. I switched to this implementation, where the cFE tracks its own task info. It really isn't much of a hack anyway.

If you think I should do it the other way, then I can spend some time to figure that out. At least this new implementation does not involve memory corruption.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+OS_TaskDelay(uint32 millisecond)
+{
+	cycles_t wakeup = sl_now() + sl_usec2cyc(millisecond * 1000);
+	sl_thd_block_timeout(0, wakeup);
```
empty lines.

gparmer replied:
wow, relative task delay APIs are broken.  Way to go cfe.  (though Posix made the same mistake).

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+	struct sl_thd *thd = sl_thd_lkup(task_id);
+	if (!thd) { return OS_ERR_INVALID_ID; }
+
+	union sched_param_union sp = {.c = {.type = SCHEDP_PRIO, .value = new_priority}};
```
defblockblah

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+{
+	if (!task_id || !task_name) return OS_INVALID_POINTER;
+
+	thdid_t i;
```
defblockblah

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+	thdid_t i;
+	for (i = 1; i < SL_MAX_NUM_THDS; i++) {
+		struct sl_thd *thd = sl_thd_lkup(i);
+		if (!thd || thd->state == SL_THD_FREE) continue;
```
empty line.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+		if (!thd || thd->state == SL_THD_FREE) continue;
+		if (strcmp(cfe_tasks[i].osal_task_prop.name, task_name) == 0) {
+			*task_id = i;
+			return OS_SUCCESS;
```
empty line

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+{
+	if (!task_prop) { return OS_INVALID_POINTER; }
+
+	struct sl_thd *thd = sl_thd_lkup(task_id);
```
def at head of blocks

gparmer replied:
OK, I'm going to stop listing these so as to not be to annoying.  You can fix this throughout easily.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+*/
+
+/*
+** OS-specific background thread implementation - waits forever for events to occur.
```
Is this a cFE comment style?  It doesn't match our typical one, or DOxygen.

Others replied:
This is from the cFE. Their comment style is all over the place...

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id == NULL || sem_name == NULL) {
```
do this and the next error check *before* taking the lock.  Always want to minimize lock hold time, while, of course, maintaining correctness.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+{
+	int32 result = OS_SUCCESS;
+
+	sl_lock_take(&mutex_data_lock);
```
Yeah, this hella deadlocks for the most trivial of interleavings.  Bummer.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+		goto exit;
+	}
+
+	sl_lock_take(&mutexes[sem_id].lock);
```
A simple solution that doesn't really scale in complexity is to just get rid of the `mutex_data_lock`, and take each mutex's lock separately.  Since we're in a static data-structure, this is likely fine.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+OS_MutSemDelete(uint32 sem_id)
+{
+	int32 result = OS_SUCCESS;
+	sl_lock_take(&mutex_data_lock);
```
empty line before

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id == NULL || sem_name == NULL) {
```
move outside of lock along with next error check

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+		}
+	}
+
+	/* The name was not found in the table,
```
```
/*
 * ...
 */
```

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+
+	sl_lock_take(&semaphore_data_lock);
+
+	if (sem_id == NULL || sem_name == NULL) {
```
move these error checks out of the lock

gparmer replied:
since these checks are repetitive, you might want to move them into a function.

gparmer commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+OS_SemaphoreDelete(struct semaphore *semaphores, uint32 max_semaphores, uint32 sem_id)
+{
+	int32 result = OS_SUCCESS;
+	sl_lock_take(&semaphore_data_lock);
```
empty line

Others commented on src/components/implementation/no_interface/cFE_booter/ostask.c:
```diff
@@ -0,0 +1,815 @@
+void
+osal_task_entry_wrapper(void *task_entry)
+{
+	((osal_task_entry)task_entry)();
```
Unfortunately, this is a cFE definition
